File: src/test/java/de/spries/fleetcommander/model/core/universe/UniverseFactoryTest.java
package de.spries.fleetcommander.model.core.universe;

import static org.hamcrest.Matchers.greaterThan;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.notNullValue;
import static org.hamcrest.Matchers.nullValue;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.mock;

import java.util.Arrays;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import org.junit.Test;

import de.spries.fleetcommander.model.core.Player;

public class UniverseFactoryTest {

	private static final Player JOHN = mock(Player.class);
	private static final Player OTHER_PLAYER = mock(Player.class);
	private static final List<Player> JOHN_ONLY = Arrays.asList(JOHN);

	@Test
	public void generatedUniverseHasMorePlanetsThanPlayers() throws Exception {
		Universe universe = UniverseFactory.generate(JOHN_ONLY);
		assertThat(universe.getPlanets().size(), greaterThan(JOHN_ONLY.size()));
	}

	@Test
	public void playerHasHomePlanet() throws Exception {
		Universe universe = UniverseFactory.generate(JOHN_ONLY);
		assertThat(universe.getHomePlanetOf(JOHN), is(notNullValue()));
	}

	@Test
	public void playersNotParticipatingInGameDontHaveAHomePlanet() throws Exception {
		Universe universe = UniverseFactory.generate(JOHN_ONLY);
		assertThat(universe.getHomePlanetOf(OTHER_PLAYER), is(nullValue()));
	}

	@Test
	public void everyPlanetHasAUniqueId() throws Exception {
		Universe universe = UniverseFactory.generate(JOHN_ONLY);
		Set<Integer> planetIds = universe.getPlanets().stream().map(p -> p.getId())
				.collect(Collectors.toSet());

		assertThat(planetIds.size(), is(universe.getPlanets().size()));
	}
}

File: src/test/kotlin/de/spries/fleetcommander/model/core/universe/UniverseFactoryTest.kt
package de.spries.fleetcommander.model.core.universe

import com.nhaarman.mockitokotlin2.mock
import de.spries.fleetcommander.model.core.Player
import org.hamcrest.Matchers.`is`
import org.hamcrest.Matchers.greaterThan
import org.hamcrest.Matchers.notNullValue
import org.hamcrest.Matchers.nullValue
import org.junit.Assert.assertThat
import org.junit.Test

class UniverseFactoryTest {

    @Test
    fun generatedUniverseHasMorePlanetsThanPlayers() {
        val universe = UniverseFactory.generate(JOHN_ONLY)
        assertThat(universe.getPlanets().size, greaterThan(JOHN_ONLY.size))
    }

    @Test
    fun playerHasHomePlanet() {
        val universe = UniverseFactory.generate(JOHN_ONLY)
        assertThat(universe.getHomePlanetOf(JOHN), `is`(notNullValue()))
    }

    @Test
    fun playersNotParticipatingInGameDontHaveAHomePlanet() {
        val universe = UniverseFactory.generate(JOHN_ONLY)
        assertThat(universe.getHomePlanetOf(OTHER_PLAYER), `is`(nullValue()))
    }

    @Test
    fun everyPlanetHasAUniqueId() {
        val universe = UniverseFactory.generate(JOHN_ONLY)
        val planetIds = universe.getPlanets().map { p -> p.id }

        assertThat(planetIds.size, `is`(universe.getPlanets().size))
    }

    companion object {

        private val JOHN = mock<Player>()
        private val OTHER_PLAYER = mock<Player>()
        private val JOHN_ONLY = listOf(JOHN)
    }
}
